src/serde.rs

```diff
@@ -67,59 +74,261 @@ pub fn skip_if_true(b: &bool) -> bool {
+        }
+        BytesOrStr::Str(s) => s,
+    };
+    let hex = hex.as_ref();
+
+    if hex.len() % 2 != 0 {
+        return Err(de::Error::invalid_length(
+            hex.len(),
+            &"even-numbered length",
+        ));
+    }
+
+    let mut bytes = Vec::with_capacity(hex.len() / 2);
+    let mut byte = 0;
+    for (i, digit) in hex.chars().enumerate() {
+        byte <<= 4;
+        byte |= match digit {
+            '0'..='9' => digit as u8 - b'0',
+            'a'..='f' => digit as u8 - b'a' + 0xa,
+            'A'..='F' => digit as u8 - b'A' + 0xA,
+            _ => {
+                return Err(de::Error::invalid_value(
+                    de::Unexpected::Char(digit),
+                    &"hex digit",
+                ));
+            }
+        };
+        if i % 2 == 1 {
+            bytes.push(byte)
+        }
+    }
+
+    bytes.try_into().map_err(|_| try_fail())
+}
+
+/// For deserializing a `Vec<u8>` from either a string of hex digits or a
+/// sequence of bytes.
+#[cfg(feature = "std")]
+pub fn de_hexstring<'de, D, B>(d: D) -> Result<B, D::Error>
 where
-    'de: 'a,
     D: Deserializer<'de>,
+    B: TryFrom<Vec<u8>>,
 {
-    use core::convert::TryInto as _;
-
-    let slice = <&[u8]>::deserialize(d)?;
-    slice.try_into().map_err(|_| {
-        <D::Error as de::Error>::invalid_length(
-            slice.len(),
-            &ExpectedByDisplay(N),
-        )
-    })
+    hex_to_bytes::<B, D::Error>(BytesOrStr::deserialize(d)?)
 }
 
-/// For deserializing an `&[[u8; N]]`.
-pub fn de_slice_of_u8_arrays<'de, 'a, D, const N: usize>(
-    d: D,
-) -> Result<&'a [[u8; N]], D::Error>
+/// For deserializing a `Vec<Vec<u8>>` from either a string of hex digits or a
+/// sequence of bytes.
+#[cfg(feature = "std")]
+pub fn de_hexstrings<'de, D, B>(d: D) -> Result<Box<[B]>, D::Error>
 where
-    'de: 'a,
     D: Deserializer<'de>,
+    B: TryFrom<Vec<u8>>,
+{
+    let hexen = Vec::<BytesOrStr>::deserialize(d)?;
+    let mut bufs = Vec::with_capacity(hexen.len());
+    for hex in hexen {
+        bufs.push(hex_to_bytes::<B, D::Error>(hex)?);
+    }
+    Ok(bufs.into_boxed_slice())
+}
+
+/// For serializing a `Vec<u8>` as a hexstring.
+#[cfg(feature = "std")]
+pub fn se_hexstring<S>(
+    bytes: &impl AsRef<[u8]>,
+    s: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let bytes = bytes.as_ref();
+    if s.is_human_readable() {
+        let mut output = String::with_capacity(bytes.len() * 2);
+        for byte in bytes {
+            let _ = write!(output, "{:02x}", byte);
+        }
+        s.serialize_str(&output)
+    } else {
+        s.serialize_bytes(bytes)
+    }
+}
+
+/// For serializing a `Vec<u8>` as a hexstring.
+#[cfg(not(feature = "std"))]
+pub fn se_hexstring<S>(
+    bytes: &impl AsRef<[u8]>,
+    s: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let bytes = bytes.as_ref();
+    // Due to a limitation in serde, we fall back to serializing just bytes
```

Comment:
As someone who isn't very familiar with Serde, I find "a limitation in serde" very vague. Could you elaborate?

Reply:
Serde requires that, when serializing a string, the entire string is buffered a priori, rather than exposing a writer-like interface like it does with sequences. Hence, I would need to allocate a buffer twice as large as `bytes` to serialize it, and I obviously can't thread an arena into this function.